libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt

```diff
@@ -0,0 +1,133 @@
+
+import io.element.android.libraries.core.coroutine.CoroutineDispatchers
+import io.element.android.libraries.matrix.api.encryption.BackupState
+import io.element.android.libraries.matrix.api.encryption.BackupUploadState
+import io.element.android.libraries.matrix.api.encryption.EnableRecoveryProgress
+import io.element.android.libraries.matrix.api.encryption.EncryptionService
+import io.element.android.libraries.matrix.api.encryption.RecoveryState
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.withContext
+import org.matrix.rustcomponents.sdk.BackupStateListener
+import org.matrix.rustcomponents.sdk.BackupSteadyStateListener
+import org.matrix.rustcomponents.sdk.Client
+import org.matrix.rustcomponents.sdk.EnableRecoveryProgressListener
+import org.matrix.rustcomponents.sdk.Encryption
+import org.matrix.rustcomponents.sdk.RecoveryStateListener
+import org.matrix.rustcomponents.sdk.BackupState as RustBackupState
+import org.matrix.rustcomponents.sdk.BackupUploadState as RustBackupUploadState
+import org.matrix.rustcomponents.sdk.EnableRecoveryProgress as RustEnableRecoveryProgress
+import org.matrix.rustcomponents.sdk.RecoveryState as RustRecoveryState
+
+internal class RustEncryptionService(
+    client: Client,
+    private val dispatchers: CoroutineDispatchers,
+) : EncryptionService {
+
+    private val service: Encryption = client.encryption()
+
+    private val backupStateMapper = BackupStateMapper()
+    private val recoveryStateMapper = RecoveryStateMapper()
+    private val enableRecoveryProgressMapper = EnableRecoveryProgressMapper()
+    private val backupUploadStateMapper = BackupUploadStateMapper()
+
+    override val backupStateStateFlow: MutableStateFlow<BackupState> = MutableStateFlow(service.backupState().let(backupStateMapper::map))
+    override val recoveryStateStateFlow: MutableStateFlow<RecoveryState> = MutableStateFlow(service.recoveryState().let(recoveryStateMapper::map))
+    override val enableRecoveryProgressStateFlow: MutableStateFlow<EnableRecoveryProgress> = MutableStateFlow(EnableRecoveryProgress.Unknown)
+    override val backupUploadStateStateFlow: MutableStateFlow<BackupUploadState> = MutableStateFlow(BackupUploadState.Unknown)
+
+    fun start() {
+        service.backupStateListener(object : BackupStateListener {
+            override fun onUpdate(status: RustBackupState) {
+                backupStateStateFlow.value = backupStateMapper.map(status)
+            }
+        })
+
+        service.recoveryStateListener(object : RecoveryStateListener {
+            override fun onUpdate(status: RustRecoveryState) {
+                recoveryStateStateFlow.value = recoveryStateMapper.map(status)
+            }
+        })
+    }
+
+    fun destroy() {
+        // No way to remove the listeners...
+        service.destroy()
+    }
+
+    override suspend fun enableBackups(): Result<Unit> = withContext(dispatchers.io) {
+        runCatching {
+            service.enableBackups()
```

Comment:
Just a question, why none of these ffi methods are `suspend`? Didn't we just get rid of all `*blocking` versions of many APIs?